### PR TITLE
WIP: Refactor replicaset configuration

### PR DESCRIPTION
**Pull request by josephholsten**
DO NOT MERGE

I'm trying to clean up the `configure_replicaset` method. This PR is just for others to review. Even when it is actually close to being well factored, it will need to be entirely rewritten to ensure backward compatibility.

EDIT: removing `configure_shards` from scope

TODO:
- [x] make ReplicasetMember sortable
- [x] make ReplicasetMember comparable (especially to default values in http://docs.mongodb.org/manual/reference/replica-configuration/)
- [x] extract chef-node to ReplicasetMember converter, describe how this might work for zookeeper, serf
- [ ] test calls to external libs, esp mongo gem
- [x] simplify error handling (just fail fast?)

### CODE REVIEW

File: libraries/mongodb.rb [hunk 1]
```diff
@@ -21,7 +21,158 @@
...
+
+    def ipaddress_host
+      "#{ipaddress}:#{mongodb_port}"
+    end
+
+    def host
+      fqdn_host
+    end
+
+    def arbiter_only
```
Comment (jspiewak) [inline libraries/mongodb.rb]: predicate method `arbiter_only?`

**Comment by jspiewak** [comment]
This will be much easier to review once #220 is merged and this can be rebased.
I will take another look then.

**Comment by josephholsten** [comment]
I've rebased this onto master. Its unit tests are passing locally, and I'm running integration tests now.

I'm also working on the chef-node service discovery extraction in a project called discoteq. The golang version proof of concept is at [`josephholsten/discoteq`](https://github.com/josephholsten/discoteq). I'll be releasing the ruby form (with standalone binary, chef recipe helper, and capistrano helper) once we're using it in production at sm.

I've also updated @simplymeasured's setup to be using the latest stable release, so this is finally a priority of mine again.

File: libraries/mongodb.rb (hunk 1 above)
Comment (josephholsten) [inline libraries/mongodb.rb]: I like it! Patch added.

File: libraries/mongodb.rb [hunk 2]
```diff
@@ -28,180 +32,112 @@ def self.configure_replicaset(node, name, members)
...
-        connection.database_names # check connection
-      end
-    rescue => e
-      Chef::Log.warn("Could not connect to database: 'localhost:#{node['mongodb']['config']['port']}', reason: #{e}")
-      return
-    end
+    connection = make_slave_connection 'localhost', node['mongodb']['port']
 
-    # Want the node originating the connection to be included in the replicaset
+    # ensure the node originating the connection is be included in the replicaset
```
Comment (jspiewak) [inline libraries/mongodb.rb]: "is be"

File: libraries/mongodb.rb [hunk 3]
```diff
@@ -28,180 +32,112 @@ def self.configure_replicaset(node, name, members)
...
-      begin
-        connection = Mongo::Connection.new(server, port, :op_timeout => 5, :slave_ok => true)
-      rescue
-        abort("Could not connect to database: '#{server}:#{port}'")
-      end
+    elsif result.fetch('errmsg', nil) =~ /(\S+) is already initiated/ ||
+          result.fetch('errmsg', nil) == 'already initialized'
+      # replicaset is initialized, requires reconfig
+
+      # retrieve host to from errmsg
```
Comment (jspiewak) [inline libraries/mongodb.rb]: "host to from"

File: libraries/replicaset_member.rb [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'json'
+
+# helper to cast any object to a boolean
+def Boolean(obj) # rubocop:disable MethodName
```
Comment (jspiewak) [inline libraries/replicaset_member.rb]: Extract to a helper class so it isn't duplicated?

**Comment by josephholsten** [comment]
@jspiewak @jamesonjlee I'd really love to get your opinion of the readability of this updated `configure_replicaset()`. Does it finally describe the business logic at a high level? Got any suggestions to further simplify?

**Comment by jamesonjlee** [comment]
I will give it a look over, does #307 need to go in first?

**Comment by jspiewak** [comment]
:chart_with_upwards_trend:

**Comment by josephholsten** [comment]
@jamesonjlee #307 is entirely orthogonal to this changeset.

But both will be required to support multiple instances on the same machine. Accessing node attributes is strewn all over the code. Notably when replicaset or sharding operations attempt to talk to the local server assume that they can talk to `localhost:{node.mongodb.port}`. How's that going to work with multiple instances, eh?

**Comment by josephholsten** [comment]
OK, this more or less reflects the API I'm going for with replicaset configuration. While cutting the length of the do-what-I-mean function (now `ReplicasetCoordinator#configure!`) is nice, the most important part is fully separating the interface for describing the replicaset members and the local service endpoint from the part that actually configures replication. Which is to say, it's compatible with a fix for #215 and #278! It also gives us a way to do #122 without insane hacks.

I've still got to flesh out the tests, but `replicaset_coordinator_spec.rb` lists all the states we need to test. Once I've got it covering all the edge cases, I'm going to backport to work with master's `configure_replicaset()` so we can ensure backward compatibility.

I have discovered that chefspec tests and regular specs really don't like each other, seemingly due to requires being processed multiple times. This really needs to be resolved, I'm not comfortable choosing between one set of tests and the other.

**Comment by jamesonjlee** [comment]
couldn't you rspec inside of chefspec? or I guess couldn't you just run chefspec (and have it run the rspec stuff).

**Comment by josephholsten** [comment]
@jamesonjlee yeah, that's exactly what's confusing me, chefspec is built on rspec.

**Comment by jamesonjlee** [comment]
at the end-of-the-day I think chefspec is probably what we will use.
